operator/src/main/java/oracle/kubernetes/operator/utils/DomainUpgradeUtils.java

```diff
@@ -0,0 +1,394 @@
+import oracle.kubernetes.weblogic.domain.model.AuxiliaryImageEnvVars;
+import oracle.kubernetes.weblogic.domain.model.Domain;
+import org.jetbrains.annotations.NotNull;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.Yaml;
+
+import static oracle.kubernetes.operator.ProcessingConstants.COMPATIBILITY_MODE;
+import static oracle.kubernetes.operator.helpers.LegalNames.toDns1123LegalName;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_DEFAULT_INIT_CONTAINER_COMMAND;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_INIT_CONTAINER_NAME_PREFIX;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_INIT_CONTAINER_WRAPPER_SCRIPT;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_TARGET_PATH;
+import static oracle.kubernetes.weblogic.domain.model.AuxiliaryImage.AUXILIARY_IMAGE_VOLUME_NAME_PREFIX;
+
+@SuppressWarnings({"Convert2MethodRef", "unchecked", "rawtypes"})
+public class DomainUpgradeUtils {
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Webhook", "Operator");
+
+  public static final String API_VERSION_V9 = "weblogic.oracle/v9";
+  public static final String API_VERSION_V8 = "weblogic.oracle/v8";
+  private volatile int containerIndex = 0;
+
+  public ConversionReviewModel readConversionReview(String resourceName) {
+    return getGsonBuilder().fromJson(resourceName, ConversionReviewModel.class);
+  }
+
+  /**
+   * Create the conversion review response.
+   * @param conversionRequest The request to be converted.
+   * @return ConversionResponse The response to the conversion request.
+   */
+  public ConversionResponse createConversionResponse(ConversionRequest conversionRequest) {
+    List<Object> convertedDomains = new ArrayList<>();
+
+    conversionRequest.getObjects()
+            .forEach(domain -> convertedDomains.add(
+                    convertDomain((Map<String,Object>) domain, conversionRequest.getDesiredAPIVersion())));
+
+    return new ConversionResponse()
+            .uid(conversionRequest.getUid())
+            .result(new Result().status("Success"))
+            .convertedObjects(convertedDomains);
+  }
+
+  public String writeConversionReview(ConversionReviewModel conversionReviewModel) {
+    return getGsonBuilder().toJson(conversionReviewModel, ConversionReviewModel.class);
+  }
+
+  /**
+   * Convert the domain to desired API version.
+   * @param domain Domain to be converted.
+   * @return Domain The converted domain.
+   */
+  public Object convertDomain(Map<String, Object> domain, String desiredAPIVersion) {
+    Map<String, Object> spec = getSpec(domain);
+    if (spec == null) {
+      return domain;
+    }
+    LOGGER.fine("Converting domain " + domain + " to " + desiredAPIVersion + " apiVersion.");
+
+    String apiVersion = (String)domain.get("apiVersion");
+    adjustAdminPortForwardingDefault(spec, apiVersion);
```

Comment:
I see transformations for the port-forwarding change and for auxiliary images... where is the change for removing the Progressing condition?

Reply:
I'll read below, but wondering if these methods will each handle both the v8 -> v9 and the v9 -> v8 conversion. I suspect those code flows will be quite different and it might read more intuitively if we made the conversion decision up front and then had specific methods.

Reply:
My understanding was that "Progressing condition" is removed during upgrade in DomainStatus.java (line 576)

```
  public DomainStatus upgrade() {
    Optional.ofNullable(conditions).ifPresent(x -> x.removeIf(cond -> cond.hasType(Progressing)));
    return this;
  }
```

Reply:
> My understanding was that "Progressing condition" is removed during upgrade in DomainStatus.java (line 576)

  public DomainStatus upgrade() {
    Optional.ofNullable(conditions).ifPresent(x -> x.removeIf(cond -> cond.hasType(Progressing)));
    return this;
  }

If your webhook removes Progressing then we don't need the upgrade() method any more. Our intention was to not have Progressing in v9; however, we were blocked by not yet having the webhook.

Reply:
ok. I will make the change in the webhook to remove the Progressing condition.